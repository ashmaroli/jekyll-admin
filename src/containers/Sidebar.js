import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router';
import _ from 'underscore';
import classnames from 'classnames';

import { capitalize } from '../utils/helpers';
import { ADMIN_PREFIX } from '../constants';

export class Sidebar extends Component {

  constructor(props){
    super(props);

    this.handleClick = this.handleClick.bind(this);
    this.state = { collapsedPanel: true };
  }

  handleClick() {
    if (this.state.collapsedPanel == true) {
      this.setState({
        collapsedPanel: false
      });
    } else {
      this.setState({
        collapsedPanel: true
      });
    }
  }

  linkTo(item, label) {
    return item.type === 'collection' ?
      `${ADMIN_PREFIX}/collections/${label}` :
      `${ADMIN_PREFIX}/${label}`;
  }

  renderPanels(item) {
    return _.map(item.panels, (panel, i) => {
      return (
        <li key={i}>
          <Link activeClassName="active" to={this.linkTo(item, panel)}>
            <i className="fa fa-book" />{capitalize(panel)}
          </Link>
        </li>
      );
    });
  }

  renderAccordionRoute(item, index) {
    const panelHeight = this.state.collapsedPanel ? 47 : (item.panels.length + 1) * 49;
    const accordionClasses = classnames({
      'accordion-label': true,
      'collapsed': this.state.collapsedPanel
    });

    return (
      <li className={accordionClasses} style={{ maxHeight: panelHeight }} key={index}>
        <a onClick={this.handleClick}>
          <i className={`fa fa-${item.icon}`} />
          {capitalize(item.label)}
          <div className="counter">{item.panels.length}</div>
          <div className="chevrons">
            <i className="fa fa-chevron-up" />
          </div>
        </a>
        <ul>
          {this.renderPanels(item)}
        </ul>
      </li>
    );
  }

  renderRoute(item, index) {
    if (item.render) {
      return (
        <li key={index}>
          <Link activeClassName="active" to={this.linkTo(item, item.link || item.label)}>
            <i className={`fa fa-${item.icon}`} />
            {capitalize(item.label)}
          </Link>
        </li>
      );
    }
  }

  renderRouteSection(section) {
    return _.map(section, (item, index) => {
      if (item.panels && item.panels.length > 0) {
        return this.renderAccordionRoute(item, index);
      } else {
        return this.renderRoute(item, index);
      }
    });
  }

  renderRoutes() {
    const { config, site } = this.props;
    const collections = _.filter(site.collections, label => label !== 'posts');

    const routes = [
      [
        {
          'label': 'pages',
          'icon': 'file-text',
          'render': site.content_pages.length > 0
        }
      ],
      [
        {
          'label': 'drafts',
          'icon': 'edit',
          'render': site.drafts.length > 0
        },
        {
          'label': 'posts',
          'icon': 'book',
          'type': 'collection',
          'render': site.posts.length > 0
        }
      ],
      [
        {
          'label': 'collections',
          'icon': 'briefcase',
          'type': 'collection',
          'panels': collections,
          'render': collections.length > 0
        }
      ],
      [
        {
          'label': 'data files',
          'icon': 'database',
          'link': 'datafiles',
          'render': site.data_files.length > 0
        },
        {
          'label': 'static files',
          'icon': 'file',
          'link': 'staticfiles',
          'render': site.static_files.length > 0
        }
      ],
      [
        {
          'label': 'templates',
          'icon': 'clone',
          'render': site.templates.length > 0
        },
        {
          'label': 'theme gem',
          'icon': 'diamond',
          'link': 'theme',
          'render': config.theme ? true : false
        }
      ],
      [
        {
          'label': 'configuration',
          'icon': 'cog',
          'render': true
        }
      ]
    ];

    return _.map(routes, (section, index) => {
      const section_size = _.map(section, entry => entry.render).filter(Boolean).length;

      if (section_size > 0) {
        return (
          <div className="route-section" key={index}>
            {this.renderRouteSection(section)}
          </div>
        );
      }
    });
  }

  render() {
    return (
      <div className="sidebar">
        <Link className="logo" to={ADMIN_PREFIX} />
        <ul className="routes">
          {this.renderRoutes()}
        </ul>
      </div>
    );
  }
}

Sidebar.propTypes = {
  site: PropTypes.object.isRequired,
  config: PropTypes.object.isRequired
};

export default Sidebar;
