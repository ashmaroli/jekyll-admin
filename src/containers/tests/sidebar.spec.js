import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import _ from 'underscore';
import { mount } from 'enzyme';
import { Sidebar } from '../Sidebar';

import { config, site, blank_site } from './fixtures';

const defaultProps = {
  config,
  site
};

const nonCollectionLinks = ['content_pages', 'data_files', 'static_files', 'configuration', 'drafts', 'posts'];

function setup(props) {
  const component = mount(<Sidebar {...props} />);

  return {
    component,
    links: component.find('.routes').find('li')
  };
}

describe('Containers::Sidebar', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Sidebar {...defaultProps} />, div);
  });

  it('should render correctly', () => {
    const { links, component } = setup(defaultProps);
    const { config } = component.props();

    const keys = _.filter(_.keys(site), key => nonCollectionLinks.includes(key));
    const collections = site.collections.length > 1 ? site.collections.length : 0;
    const theme = config.theme ? 1 : 0;

    const actual = links.length;
    const expected = keys.length + collections + site.templates.length + theme + 1; // the link to /configuration

    expect(actual).toEqual(expected);

    const tree = renderer
      .create(<Sidebar {...defaultProps} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render collapsible list-item for collections', () => {
    const { component, links } = setup(defaultProps);
    const listItem = links.find('.accordion-label');
    expect(listItem.text()).toContain('Collections');
    expect(component.state('collapsedPanel')).toBe(true);

    listItem.find('a').first().simulate('click');
    expect(component.state('collapsedPanel')).toBe(false);
    listItem.find('a').first().simulate('click');
    expect(component.state('collapsedPanel')).toBe(true);
  });

  it('should render fine for a "blank" Jekyll site', () => {
    const minimal_config = { gems: ['jekyll-admin'] };
    const { links } = setup({
      site: blank_site,
      config: minimal_config
    });
    expect(links.length).toEqual(1); // the link to /Configuration
  });
});
