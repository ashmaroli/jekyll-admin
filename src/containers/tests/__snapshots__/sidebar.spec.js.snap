// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Containers::Sidebar should render correctly 1`] = `
<div
  className="sidebar"
>
  <a
    className="logo"
    onClick={[Function]}
    style={Object {}}
  />
  <ul
    className="routes"
  >
    <div
      className="route-section"
    >
      <li>
        <a
          onClick={[Function]}
          style={Object {}}
        >
          <i
            className="fa fa-file-text"
          />
          Pages
        </a>
      </li>
    </div>
    <div
      className="route-section"
    >
      <li>
        <a
          onClick={[Function]}
          style={Object {}}
        >
          <i
            className="fa fa-edit"
          />
          Drafts
        </a>
      </li>
      <li>
        <a
          onClick={[Function]}
          style={Object {}}
        >
          <i
            className="fa fa-book"
          />
          Posts
        </a>
      </li>
    </div>
    <div
      className="route-section"
    >
      <li
        className="accordion-label collapsed"
        style={
          Object {
            "maxHeight": 47,
          }
        }
      >
        <a
          onClick={[Function]}
        >
          <i
            className="fa fa-briefcase"
          />
          Collections
          <div
            className="counter"
          >
            1
          </div>
          <div
            className="chevrons"
          >
            <i
              className="fa fa-chevron-up"
            />
          </div>
        </a>
        <ul>
          <li>
            <a
              onClick={[Function]}
              style={Object {}}
            >
              <i
                className="fa fa-book"
              />
              Puppies
            </a>
          </li>
        </ul>
      </li>
    </div>
    <div
      className="route-section"
    >
      <li>
        <a
          onClick={[Function]}
          style={Object {}}
        >
          <i
            className="fa fa-database"
          />
          Data files
        </a>
      </li>
      <li>
        <a
          onClick={[Function]}
          style={Object {}}
        >
          <i
            className="fa fa-file"
          />
          Static files
        </a>
      </li>
    </div>
    <div
      className="route-section"
    >
      <li>
        <a
          onClick={[Function]}
          style={Object {}}
        >
          <i
            className="fa fa-clone"
          />
          Templates
        </a>
      </li>
      <li>
        <a
          onClick={[Function]}
          style={Object {}}
        >
          <i
            className="fa fa-diamond"
          />
          Theme gem
        </a>
      </li>
    </div>
    <div
      className="route-section"
    >
      <li>
        <a
          onClick={[Function]}
          style={Object {}}
        >
          <i
            className="fa fa-cog"
          />
          Configuration
        </a>
      </li>
    </div>
  </ul>
</div>
`;
